recipes/read-excel/1.1.4.1/test_package/CMakeLists.txt

```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.1)
+
+set( CMAKE_CXX_STANDARD 14 )
+
+set( CMAKE_CXX_STANDARD_REQUIRED ON )
+
+project(read-excel.test)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+include_directories(${CONAN_INCLUDE_DIRS_READ-EXCEL})
```

Comment:
Your conanfile recipe does not describe the same cmake targets as the cmake install script of read-excel intstalls.
See https://github.com/igormironchik/read-excel/blob/a86edb9c4a88401efed875a1578cf630536438ec/CMakeLists.txt#L41-L51

I think you need to do something similar as
https://github.com/conan-io/conan-center-index/blob/master/recipes/leptonica/all/conanfile.py#L184-L187 , 
https://github.com/conan-io/conan-center-index/blob/master/recipes/leptonica/all/conanfile.py#L189-L199
and
https://github.com/conan-io/conan-center-index/blob/master/recipes/leptonica/all/conanfile.py#L216-L218

(But creating the `excel` target instead)

Then, you can do the following instead:
```cmake
add_executable(test_package ...)
target_link_libraries(test_package PRIVATE excel)
```
It looks like the install cmake script at https://github.com/igormironchik/read-excel/ does not install excel in a namespace.

Reply:
@madebr Yes, and what? I don't build `read-excel` in Conan recipe. This is just a header only library. We don't need to use CMake here at all. I don't install the library with CMake, but I use in recipe:

```
def package(self):
        self.copy("COPYING", src=self._source_subfolder, dst="licenses")
        self.copy("*.hpp", src=os.path.join(self._source_subfolder, "excel"), dst=os.path.join("include", "read-excel"))
```

That's all!!! I don't provide CMake targets at all!!!

In native installation I use `excel` as name, but in GitHub project names `read-excel`. To avoid possible overlays I changed include directory to `read-excel`. In Conan's scale it's better, much more better.

Really, I didn't understand what you want.

Reply:
What I want to do is make the following possible:
**conanfile.txt**
```
[requires]
read-excel/1.1.4.1
[generators]
cmake_find_package
```
**CMakeLists.txt**
```cmake
...
find_package(excel REQUIRED)
target_link_libraries(my_executable PRIVATE excel)
```

Reply:
# Full example

**conanfile.txt**

```
[requires]
read-excel/1.1.4.1

[generators]
cmake
```

**CMakeLists.txt**

```
cmake_minimum_required(VERSION 3.1)

set( CMAKE_CXX_STANDARD 14 )

set( CMAKE_CXX_STANDARD_REQUIRED ON )

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup()

project(read-excel.test)

add_executable(${PROJECT_NAME} main.cpp)
```

**main.cpp**

```
#include <read-excel/book.hpp>
#include <read-excel/exceptions.hpp>
#include <read-excel/compoundfile/compoundfile_exceptions.hpp>

#include <iostream>

int main()
{
    try {
        Excel::Book book( "../../sample.xls" );

        Excel::Sheet * sheet = book.sheet( 0 );

        if(sheet->cell( 0, 0 ).getString() == L"This is a string.")
            return 0;

        return 1;
    }
    catch(const Excel::Exception &)
    {
        return 1;
    }
    catch(const CompoundFile::Exception &)
    {
        return 1;
    }
    catch(const std::exception &)
    {
        return 1;
    }
}
```

Reply:
You don't need to `find_package()`, you install it with `Conan`.

Reply:
This also allows to seamlessly switch between system headers (=headers provided by linux system packages) and conan packages.
That way, some project can be built without change on Windows/Linux/Macos using conan and on Linux/Macos using the system package manager (dnf/yum/apt/brew).

Reply:
Agree, it's nice candy to have ability to use CMake's `find_package()`. But is it so extremely needed in this case with `read-excel`? I understood your position, I can make everything you requested, it's not a problem. But developers are so lazy :)

Please request blocking changes on this PR.

Reply:
Can I ask you to write a good road map for Conan recipes? What the recipe should provide for header-only library, for library with tools, for usual libraries? And it would be very good if the requirements for recipes will be the same for every recipe in Conan. Really guys, one PR goes with one requirements, another with another...

Reply:
I understand.

I propose the following:
Modify the `CMakeLists.txt` of your library as following:

```cmake
add_library(read-excel INTERFACE)
add_libraries(read-excel::read-excel ALIAS read-excel)
target_include_directories(read-excel
    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
)
install(TARGETS read-excel EXPORT read-excel-exports)
install(EXPORT read-excel-exports DESTINATION "lib/cmake/read-excel/read-excel-targets.cmake" NAMESPACE read-excel::)
```
and inside your cmake `findread-excel.cmake.in`, do
```cmake
include("${CMAKE_CURRENT_LIST_DIR}/read-excel-targets.cmake")
```

Then, your library can be used as a subdirectory (submodule) and as an installed package seamlessly (when using `cmake_find_package`)

e.g.

```cmake
option(USE_READ_EXCEL_SUBMODULE)
if(USE_READ_EXCEL_SUBMODULE)
    add_subdirectory(thirdparty/read-excel)
else()
    find_package(read-excel REQUIRED)
endif()
add_executable(my_exe my_code.cpp)
target_link_libraries(my_exe PRIVATE read-excel::read-excel)
```

Reply:
> Can I ask you to write a good road map for Conan recipes? What the recipe should provide for header-only library, for library with tools, for usual libraries? And it would be very good if the requirements for recipes will be the same for every recipe in Conan. Really guys, one PR goes with one requirements, another with another...

I feel you. Sometimes I feel too that conan/cci wants to do too much or have too much control.
What I wrote in my previous comment will only work for cmake consumers but will stop when somebody is using meson or pure visual studio projects.
`conan`'s cmake_find_package` generator currently only created namespaced targets, that's why I suggested using them..

Also, `cmake_find_package` and `cmake_find_package_multi` will go away in conan 2.0. So a lot of what I write here will be old news next year.